### PR TITLE
fix #7417 feat(nimbus): display analysis errors on Results UI

### LINKED ISSUE #7417: Nimbus UI improvements to accomodate updated metrics metadata from Jetstream
This is a follow-up to [https://mozilla-hub.atlassian.net/browse/CIRRUS-781](https://mozilla-hub.atlassian.net/browse/CIRRUS-781|smart-link) for updating the UI side using new fields in the Jetstream metadata for metrics. We should be able to create a more complete display using the new status information created as part of the referenced task.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/EXP-2522)

### PR DESCRIPTION
**Pull request by mikewilli**
Because

* Jetstream now makes analysis error logs available
* there is no indication in Experimenter of why results may be missing

This commit

* retrieves errors from Jetstream
* determines if errors should be displayed
  * for Experiment level errors: display errors at top of experiment if they exist
  * for Metric level errors: if the metric has errors but does not have results
* creates an error alert component to display when appropriate

**Screenshots:**
Experiment error
image
Metric error
image

### CODE REVIEW

File: app/experimenter/jetstream/client.py [hunk 1]
```diff
@@ -171,4 +180,19 @@ def get_experiment_data(experiment):
...
         experiment_data[window] = transformed_data
 
+    errors_by_metric = {}
+    errors_experiment_overall = []
+    if experiment_errors is not None:
+        for err in experiment_errors:
+            if "metric" in err and err.get("metric") is not None:
+                if err.get("metric") not in errors_by_metric:
+                    errors_by_metric[err.get("metric")] = []
+                errors_by_metric[err.get("metric")].append(err)
```
Comment (jaredlockhart) [inline app/experimenter/jetstream/client.py:190]: Doing the same look up 4 times here, could've pulled it out into

```py
err_metric = err.get('metric')
```

Not critical just a suggestion.

File: app/experimenter/nimbus-ui/src/components/AppLayoutSidebarLaunched/index.test.tsx [hunk 2]
```diff
@@ -173,7 +175,7 @@ describe("AppLayoutSidebarLaunched", () => {
       });
     });
 
-    it("when complete and page does not require full analaysis data, has expected results page items in side bar", () => {
+    it("when complete and page does not require full analysis data, has expected results page items in side bar", () => {
```
Comment (jaredlockhart) [inline app/experimenter/nimbus-ui/src/components/AppLayoutSidebarLaunched/index.test.tsx:178]: 🙏

File: app/experimenter/nimbus-ui/src/components/AppLayoutSidebarLaunched/index.tsx [hunk 3]
```diff
@@ -212,8 +212,8 @@ export const AppLayoutSidebarLaunched = ({
                     <>
                       Could not get visualization data. Please contact data
                       science in{" "}
-                      <LinkExternal href="https://mozilla.slack.com/archives/C0149JH7C1M">
-                        #cirrus
+                      <LinkExternal href="https://mozilla.slack.com/archives/CF94YGE03">
+                        #ask-experimenter
```
Comment (jaredlockhart) [inline app/experimenter/nimbus-ui/src/components/AppLayoutSidebarLaunched/index.tsx:216]: 💯

File: app/experimenter/nimbus-ui/src/components/PageResults/TableMetricCount/MetricHeader.tsx [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  metricType = METRIC_TYPE.DEFAULT_SECONDARY,
+}: MetricHeaderProps) => {
+  const {
+    analysis: { metadata },
+  } = useContext(ResultsContext);
+
+  const outcomeName =
+    metadata?.metrics[outcomeSlug]?.friendly_name || outcomeDefaultName;
+  const outcomeDescription =
+    metadata?.metrics[outcomeSlug]?.description || undefined;
```
Comment (jaredlockhart) [inline app/experimenter/nimbus-ui/src/components/PageResults/TableMetricCount/MetricHeader.tsx:35]: Why `undefined` here and not `null`? Just curious.

File: app/experimenter/nimbus-ui/src/components/PageResults/TableMetricCount/index.tsx [hunk 5]
```diff
@@ -71,37 +70,14 @@ const TableMetricCount = ({
   );
   const outcomeName =
     metadata?.metrics[outcomeSlug]?.friendly_name || outcomeDefaultName;
-  const outcomeDescription =
-    metadata?.metrics[outcomeSlug]?.description || undefined;
```
Comment (jaredlockhart) [inline app/experimenter/nimbus-ui/src/components/PageResults/TableMetricCount/index.tsx:75]: Oh I see that's what was already here, just moved.

File: app/experimenter/nimbus-ui/src/components/PageResults/index.tsx [hunk 6]
```diff
@@ -108,46 +117,159 @@ const PageResults: React.FunctionComponent<RouteComponentProps> = () => {
...
+                          {metricHeader}
+                          <p>
+                            <i>No results available for metric.</i>
+                          </p>
+                        </>
+                      );
+                    }
+                  }
+                });
+              })}
```
Comment (jaredlockhart) [inline app/experimenter/nimbus-ui/src/components/PageResults/index.tsx:211]: If you have to do this whole same block for primary and secondary outcomes, couldn't you just pull it out into a function?

File: app/experimenter/nimbus-ui/src/lib/visualization/mocks.tsx [hunk 7]
```diff
@@ -1739,3 +1806,727 @@ export const mockIncompleteAnalysis = (modifications = {}) =>
...
+                significance: { overall: { "1": "positive" }, weekly: {} },
+              },
+            },
+            search_metrics: {
+              search_count: TREATMENT_NEGATIVE,
+            },
+          },
+        },
+      },
+    },
```
Comment (jaredlockhart) [inline app/experimenter/nimbus-ui/src/lib/visualization/mocks.tsx:2530]: Mmmmmmm look at all that test data

File: app/experimenter/nimbus-ui/src/lib/visualization/types.ts [hunk 8]
```diff
@@ -9,13 +9,35 @@ export interface AnalysisData {
...
+  message: string | null;
+  filename: string | null;
+  exception: string | null;
+  func_name: string | null;
+  log_level: string | null;
+  statistic: string | null;
+  timestamp: string | null;
+  experiment: string | null;
+  exception_type: string | null;
+}
```
Comment (jaredlockhart) [inline app/experimenter/nimbus-ui/src/lib/visualization/types.ts:39]: Oh yeah interesting, I forgot we have to maintain this by hand.  Would be really nice if the entire structure was described on the backend using Dataclasses and we could auto generate all this.

**Review by jaredlockhart — APPROVED** [review]
Okay tested locally, everything looks and works great, nicely backwards compatible.  Couple suggestions about the code but nothing huge.  Thanks @mikewilli this will make the results pages clearer and more consistent and informative 🎉 🎉 🎉 🎉

File: app/experimenter/nimbus-ui/src/components/PageResults/TableMetricCount/MetricHeader.tsx (hunk 4 above)
Comment (mikewilli) [reply app/experimenter/nimbus-ui/src/components/PageResults/TableMetricCount/MetricHeader.tsx:35]: As you noticed, I just refactored existing code and this moved along, but the important thing is that it's falsy if there is no `description` value. Personally I think `undefined` makes sense here, but either would work.

File: app/experimenter/jetstream/client.py [hunk 9]
```diff
@@ -171,4 +180,20 @@ def get_experiment_data(experiment):
...
 
+    errors_by_metric = {}
+    errors_experiment_overall = []
+    if experiment_errors is not None:
+        for err in experiment_errors:
+            metric_slug = err.get("metric")
+            if "metric" in err and metric_slug is not None:
+                if metric_slug not in errors_by_metric:
+                    errors_by_metric[metric_slug] = []
+                errors_by_metric[metric_slug].append(err)
```
Comment (jaredlockhart) [inline app/experimenter/jetstream/client.py:191]: 🙏

File: app/experimenter/nimbus-ui/src/components/PageResults/index.tsx [hunk 10]
```diff
@@ -59,6 +61,62 @@ const PageResults: React.FunctionComponent<RouteComponentProps> = () => {
 
   const { external_config: externalConfig } = analysis.metadata || {};
 
+  const getErrorsForOutcomes = (
```
Comment (jaredlockhart) [inline app/experimenter/nimbus-ui/src/components/PageResults/index.tsx:64]: 👍 👏 💯

**Review by jaredlockhart — APPROVED** [review]
Yep yup looks good, great job @mikewilli 🎉

**Pull request merged**